### PR TITLE
Add journald receiver

### PR DESCRIPTION
**Pull request by luckyj5**
Description: Add Journald receiver
Link to tracking Issue: #2332

Testing: Unit tests are included

Documentation: Journald receiver README.md is included

### CODE REVIEW

**Review by djaglowski — COMMENTED** [review]
The new module looks fine, with a few comments to address.

However, I am not clear on the implications of adding journalctl to the dockerfile. Is this best approach? Is it wise to manage such a large set of specific files in this way? Are we sure this is the right set of files? Will this be a fragile pattern? At a minimum, I think we need a test that will fail if journald cannot run due to changes in the set of files that must be copied into the image.

We also have a challenge with the windows tests failing. `opentelemetry-log-collection` only builds the journald operator for linux builds. I would think we want a way to include this receiver only in linux builds, but I don't think we have such a mechanism in place for the collector. Any ideas on the best approach here?

**Comment by luckyj5** [comment]
> The new module looks fine, with a few comments to address.
> 
> However, I am not clear on the implications of adding journalctl to the dockerfile. Is this best approach? Is it wise to manage such a large set of specific files in this way? Are we sure this is the right set of files? Will this be a fragile pattern? At a minimum, I think we need a test that will fail if journald cannot run due to changes in the set of files that must be copied into the image.
> 
> We also have a challenge with the windows tests failing. `opentelemetry-log-collection` only builds the journald operator for linux builds. I would think we want a way to include this receiver only in linux builds, but I don't think we have such a mechanism in place for the collector. Any ideas on the best approach here?

@djaglowski Thanks. PR is updated per your comments. 
Regarding best approach for adding journalctl to the dockerfile, its added based on the discussion here https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/2332. 
 
 We'll look into it a bit more and share our thoughts. Thanks!

**Comment by gillg** [comment]
I like the global idea of this receiver I was searching for it some weeks ago and I fallbacked on syslog. Moreover reuse the sideproject for logs collection seems perfect for the consistency.
But I'm a little bit affraid about the difficult to maintain the docker image with journalctl installed by this way... We could have broken builds if the debian version change, incompatibilities between amd64 / arm64 builds, cross OS difficulties and so on.

Maybe an alternative could be to propose install systemd-journal-remote to the user like we need to install rsyslog to use syslog receiver ?
In this case instead of read local files with journalctl we bind an HTTP endpoint to receive the journal content in a pre-translated format (json probably).

**Comment by rockb1017** [comment]
First, I would like to know how to not include journald receiver when building it for non-linux OS because it will only work for linux. Please let me know if you have a preferred way for it.

**Comment by rockb1017** [comment]
> But I'm a little bit affraid about the difficult to maintain the docker image with journalctl installed by this way... We could have broken builds if the debian version change, incompatibilities between amd64 / arm64 builds, cross OS difficulties and so on.
image
it has been updated 6 months ago and no changes since then. 

> In this case instead of read local files with journalctl we bind an HTTP endpoint to receive the journal content in a pre-translated format (json probably).
what is sending journald logs to HTTP endpoint? I am not fan of requiring any operation on users host side to make it work.

**Comment by djaglowski** [comment]
@rockb1017 I think we may want to do the following to the log-collection library: https://github.com/open-telemetry/opentelemetry-log-collection/pull/267/files

This should allow all operating systems to compile and pass tests, while still omitting the journald operator from non-linux builds. The result, in theory, is that users who try to use the journald receiver will get an error from the log-collection library, along the lines of "unknown operator 'journald'". (If we go this route, we may want to capture the error and wrap it in something clearer to the user.)

**Comment by rockb1017** [comment]
@djaglowski Thanks for replying. I tried something along that approach but build still fails with this. `receiver/journaldreceiver/journald.go:68:14: undefined: journald.NewJournaldInputConfig`
it is because in `internal/components/components.go`, it calls `journaldreceiver.NewFactory()`

**Comment by djaglowski** [comment]
@rockb1017 I have opened an [issue](https://github.com/open-telemetry/opentelemetry-collector-builder/issues/71) to discuss excluding components such as this, based on OS.

If this is not implemented, or not soon enough, I think the next best option may be to write a full "dummy" implementation for non-linux builds. Something like:
```go
// build !linux

package journaldreceiver

import (
	"go.opentelemetry.io/collector/component"
	"go.opentelemetry.io/collector/config"
	"gopkg.in/yaml.v2"
)

const typeStr = "journald"

// NewFactory creates a dummy factory.
func NewFactory() component.ReceiverFactory {
	return receiverhelper.NewFactory(
		typeStr,
		createDefaultConfig,
		receiverhelper.WithLogs(createLogsReceiver))
}

func createDefaultConfig() config.Receiver {
	return nil // or whatever dummy is necessary

func createLogsReceiver(
	_ context.Context,
	_ component.ReceiverCreateSettings,
	_ config.Receiver,
	_ consumer.Metrics,
) (component.MetricsReceiver, error) {
	return nil, fmt.Errorf("'journald' only supported on linux")
}
```

**Comment by gillg** [comment]
Hello again.
I bring some new information with my strategy to avoid OS dependancies on OTEL collector build and docker image.
If you want expose journald in a readable format, journalctl is the way localy and journald-gateway is the way remotely.

My idea is :
* We use a sidecar pattern container (called "journald-gateway" as exemple) built with something like that :
```
FROM debian:stable-slim

RUN apt update && apt install -y systemd-journal-remote && apt clean
# For RHEL : yum install systemd-journal-gateway

EXPOSE 19531
 
# Doc https://www.freedesktop.org/software/systemd/man/systemd-journal-gatewayd.service.html
ENTRYPOINT ["/lib/systemd/systemd-journal-gatewayd"]
# For RHEL : /usr/lib/systemd/systemd-journal-gatewayd

CMD ["--directory=/host/var/log/journal"]
```
* We create it on host as sidecar with `docker run -v /var/log/journal/:/host/var/log/journal/ -ti journald-gateway`
* Otel container can reach it with http://localhost:19531/entries?follow and http://localhost:19531/machine to build resources attributes eventualy
* On otel collector we use a config like : 
```yaml
receiver:
  journald:
    gateway:
      endpoint: http://journald-gateway:19531/
      filters:  # can be use to filter the /entries http route
```

All documentation here https://www.freedesktop.org/software/systemd/man/systemd-journal-gatewayd.service.html#Supported%20URLs
We can eventualy have only one collector for multiple nodes if we don't want the sidecar pattern.

Probably, this change a lot the actual code but ot the goal, and don't disturb otel-collector.

File: cmd/otelcontribcol/Dockerfile [hunk 1]
```diff
@@ -1,15 +1,25 @@
...
 RUN apk --update add ca-certificates
 
+ARG USER_UID=10001
+# github.com/containers/podman/v2 requires these directories to be present
+# and be writable. When this is not satisfied, the library raises a fatal
+# error and crashes the collector.
+# This is required by the podmanreceiver.
+RUN mkdir -p /rundir
+RUN mkdir -p /tmp
+RUN chown -R ${USER_UID}:${USER_UID} /rundir
```
Comment (gillg) [inline cmd/otelcontribcol/Dockerfile:11]: As already said, work on Dockerfile and introduce kind of complexity will probably introduce issues between containers runtimes and OS.
If tomorrow we provide an image for windows container (PR already exists) as an exemple, these steps will not be compatible, while just copying otelcontribcol is.

Comment (rockb1017) [reply cmd/otelcontribcol/Dockerfile:11]: the dockerfile for windows would be a separate file yea? so it would affect the image for windows.

Comment (luckyj5) [reply cmd/otelcontribcol/Dockerfile:11]: Hey @gillg thanks. Just to clarify are you referring line 8-11 in the Dockerfile or changes related to journalctl?

**Comment by rockb1017** [comment]
@gillg 
Thanks for looking into it and providing alternatives. 
I think it makes deployment complicated. in the k8s environment, all will be wrapped in a helm chart, so it doesn't matter much but for any other use case, it makes it complicated to collect journald logs. 
Also when it is consuming from gateway, how would it track offset for checkpointing? is it possible to make sure it consumes journald logs without data loss and data duplication when pod gets restarted?

**Comment by gillg** [comment]
> Also when it is consuming from gateway, how would it track offset for checkpointing? is it possible to make sure it consumes journald logs without data loss and data duplication when pod gets restarted?

Good point, you can use Range query parameter to start to a specific offset, you can put a cursor reference in a file or shared parameter. How do you deal it with journalctl ? Parameters are pretty the same.
And about dedup yes I don't have solution, but seems similar with journalctl if you start 2 container mounted on journal folder, no ?

File: cmd/otelcontribcol/Dockerfile (hunk 1 above)
Comment (gillg) [reply cmd/otelcontribcol/Dockerfile:11]: Not 100% sure if windows will be a different file. The image can be just an ARG and buildkit can create all flavors.
And my comment is pretty general but just trigger here last changes related to adaptations made in podman context.

**Comment by rockb1017** [comment]
@gillg oh there is no checkpoint mechanism with journald input operator. I assumed there is.

**Comment by djaglowski** [comment]
> @gillg oh there is no checkpoint mechanism with journald input operator. I assumed there is.

Actually, there is. See references to [cursor](https://github.com/open-telemetry/opentelemetry-log-collection/blob/bef28006119df702e94e7271bf59a4ea8101ec10/operator/builtin/input/journald/journald.go#L107)

**Comment by rockb1017** [comment]
@djaglowski Awesome! yes it does have it. i missed that part. so we just need to enable storage extension just like filelog, yea?

**Comment by djaglowski** [comment]
> so we just need to enable storage extension just like filelog, yea?

Yes, it should work exactly the same.

**Comment by djaglowski** [comment]
Please see comment [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/5160#issuecomment-920290933) about using a stub receiver for non-linux OS.

**Comment by djaglowski** [comment]
Also, the `main` branch was broken this morning, but was just fixed. Pulling in the update should resolve many of these failures.

**Comment by tigrannajaryan** [comment]
Please make sure the build passes.
```
make: *** [Makefile.Common:78: checklicense] Error 1
make: *** Waiting for unfinished jobs....
checkdoc --project-path /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib --component-rel-path internal/components/components.go --module-name github.com/open-telemetry/opentelemetry-collector-contrib
addlicense FAILED => add License errors:

./receiver/journaldreceiver/journald_nonlinux.go

Use 'make addlicense' to fix this.
Error: Process completed with exit code 2.
```

**Comment by luckyj5** [comment]
Thanks all for your inputs and suggestions. PR has been updated with the same.

File: receiver/journaldreceiver/README.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
+## `Journald Receiver`
+
+Parses Journald events using using the [opentelemetry-log-collection](https://github.com/open-telemetry/opentelemetry-log-collection) library.
```
Comment (djaglowski) [inline receiver/journaldreceiver/README.md:3]: Please rephrase as a complete sentence and remove the duplicate word.

File: receiver/journaldreceiver/README.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+## Configuration
+
+| Field                  | Default          | Description                                                                                                        |
+| ---                    | ---              | ---                                                                                                                |
+| `directory`            | /run/log/journal | A directory containing journal files to read entries from     |
```
Comment (djaglowski) [inline receiver/journaldreceiver/README.md:9]: The underlying operator will by default read from either `/run/log/journal` or `/run/journal`, whichever is present. Can you please note both here?

[journald_input docs](https://github.com/open-telemetry/opentelemetry-log-collection/blob/main/docs/operators/journald_input.md)

File: receiver/journaldreceiver/coverage.out [hunk 4]
```diff
@@ -0,0 +1 @@
+mode: atomic
```
Comment (djaglowski) [inline receiver/journaldreceiver/coverage.out:1]: I think this file is supposed to be excluded?

File: receiver/journaldreceiver/journald_nonlinux.go [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+func createLogsReceiver(
+	_ context.Context,
+	params component.ReceiverCreateSettings,
+	cfg config.Receiver,
+	consumer consumer.Logs,
+) (component.LogsReceiver, error) {
+
+	return nil, fmt.Errorf("journald is only supported on linux")
```
Comment (djaglowski) [inline receiver/journaldreceiver/journald_nonlinux.go:64]: 👍

File: receiver/journaldreceiver/README.md (hunk 2 above)
Comment (luckyj5) [reply receiver/journaldreceiver/README.md]: Done. Let me know if anything specific needs to be added. Thanks!

File: receiver/journaldreceiver/coverage.out (hunk 4 above)
Comment (luckyj5) [reply receiver/journaldreceiver/coverage.out]: Yes, thank you for pointing this. Removed it

**Review by djaglowski — COMMENTED** [review]
The receiver code looks good to me, but there are two open questions in my mind:
1. Are the changes to the Dockerfile acceptable? Should we consider adding a dedicated Dockerfile that enables this receiver? I'm not sure the if the implementation of that approach would be reasonable either. @tigrannajaryan, do you have an opinion on this?
2. Can we have an integration test, and/or test(s) in the testbed? Perhaps take inspiration from the [containerized integration tests in other receivers](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/nginxreceiver/integration_test.go#L47:L67).

**Comment by jpkrohling** [comment]
>I would suggest if we can remove dockerfile changes from this PR and make it clear in the docs that this receiver depends on journald to be available on the system.

+1, I don't think journald is even recommended to exist in containers...

**Comment by dmitryax** [comment]
> @dmitryax I assume for the use case we have for journald (to be added to Helm Chart) this is not a problem? (I don't know, just want to make sure).

@tigrannajaryan for now we can enhance docker image in our own distro since we do want to use journald by default

**Review by djaglowski — APPROVED** [review]
LGTM

**Review by tigrannajaryan — APPROVED** [review]

**Pull request merged**